src/graph/entities/attribute_set.c

```diff
@@ -149,6 +143,41 @@ void AttributeSet_Add
 		size_t n = ATTRIBUTESET_BYTE_SIZE(_set);
 		_set = rm_realloc(_set, n);
 	}
+	return _set;
+}
+
+// adds an attribute to the set without cloning the SIvalue
+void AttributeSet_AddNoClone
+(
+	AttributeSet *set,     // set to update
+	Attribute_ID attr_id,  // attribute identifier
+	SIValue value          // attribute value
+) {
+	// validate value type
+	// value must be a valid property type
+	ASSERT(SI_TYPE(value) & SI_VALID_PROPERTY_VALUE);
+	AttributeSet _set = AttributeSet_AddPrepare(set, attr_id);
+
+	// set attribute
+	Attribute *attr = _set->attributes + _set->attr_count - 1;
+	attr->id = attr_id;
+	attr->value = value;
+
+	// update pointer
+	*set = _set;
+}
+
+// adds an attribute to the set
+void AttributeSet_Add
+(
+	AttributeSet *set,     // set to update
+	Attribute_ID attr_id,  // attribute identifier
+	SIValue value          // attribute value
+) {
+	// validate value type
+	// value must be a valid property type
+	ASSERT(SI_TYPE(value) & SI_VALID_PROPERTY_VALUE);
```

Comment:
this is common to both add function why it is not in the prepare?

Reply:
It is just an assertion on the value. The prepare function does some actual logic.